In [31]:
import numpy as np
import pandas as pd

from sklearn.model_selection import train_test_split
from sklearn import preprocessing
from sklearn.linear_model import LinearRegression
from sklearn import metrics
from sklearn.model_selection import cross_val_score

# Question2: Debt Prediction

In [17]:
df  = pd.read_csv("https://github.com/mpourhoma/CS4661/raw/master/Credit.csv")

In [18]:
feature_cols = ['Income','Limit','Rating','Cards','Age','Education','Married']

X = df[feature_cols] 
print(X.shape)
print(X.head)

y = df['Balance']
print(y.shape)

(400, 7)
<bound method NDFrame.head of       Income  Limit  Rating  Cards  Age  Education  Married
0     14.891   3606     283      2   34         11        1
1    106.025   6645     483      3   82         15        1
2    104.593   7075     514      4   71         11        0
3    148.924   9504     681      3   36         11        0
4     55.882   4897     357      2   68         16        1
..       ...    ...     ...    ...  ...        ...      ...
395   12.096   4100     307      3   32         13        1
396   13.364   3838     296      5   65         17        0
397   57.872   4171     321      5   67         12        1
398   37.728   2525     192      1   44         13        1
399   18.701   5524     415      5   64          7        0

[400 rows x 7 columns]>
(400,)


In [22]:

X = preprocessing.scale(X)

In [23]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.24, random_state=9)

In [24]:
my_linreg = LinearRegression()

my_linreg.fit(X_train, y_train)

LinearRegression()

In [25]:
print(my_linreg.intercept_)

print(my_linreg.coef_)

520.0391034603759
[-264.69553256  360.5601294   252.73642988   24.91707959  -17.23897686
   11.65810466   -9.82859046]


$$y = 520.039 + -264.695 \times Income  + 360.560 \times Limit  +  252.736 \times Rating  +    24.917 \times Cards  +   -17.238 \times Age  +   11.658 \times Education  +    -9.828 \times Married$$

### It seems that the Limit has the greatest effect on the balance with married status being the least

In [26]:
y_prediction = my_linreg.predict(X_test)

print(y_prediction)

[ 316.89813114  677.5093907   489.03669272  -49.82989971  706.42913216
  864.96954115  897.65752581  108.20297732 -213.73474279  834.94198677
   11.05109139 -227.34447869  472.01450822    4.54554458 -228.17349827
  956.21537845  896.91570524  821.97824267   18.37144232  857.31517013
 1028.30326279  699.22168497 1176.9426756   666.28578465  643.94064891
  690.63628696  589.35273903  -29.37237394  575.15444917  406.78468032
  845.81453327  828.30862904   82.2376206   953.40592514 -184.74828227
  525.84335889 1045.19205022  535.04104472  107.26753118 -172.8968111
  497.84037664 1155.35533939  429.08674601  403.49872552  143.72087313
  764.78737012  431.24114308 1304.79656588  475.00132057   95.75404622
 -215.69000296    5.62900657  313.88797405  859.9328932   792.27147016
 1029.90210771 1500.40180297 1078.48617624  713.37981473 1072.97948262
  -64.12187718  488.60063387   92.3601712   440.414465    277.13284667
 1075.41508762  258.10181743  773.89657462  278.12447916  409.51029492
  335.5

In [28]:
mse = metrics.mean_squared_error(y_test, y_prediction)

In [29]:
rmse = np.sqrt(mse)

print(rmse)

143.24649739642265


## 10-fold Cross-Validation

In [37]:
feature_cols = ['Income','Limit','Rating','Cards','Age','Education','Married']

X = df[feature_cols] 
print(X.shape)
print(X.head)

y = df['Balance']
print(y.shape)

X = preprocessing.scale(X)

(400, 7)
<bound method NDFrame.head of       Income  Limit  Rating  Cards  Age  Education  Married
0     14.891   3606     283      2   34         11        1
1    106.025   6645     483      3   82         15        1
2    104.593   7075     514      4   71         11        0
3    148.924   9504     681      3   36         11        0
4     55.882   4897     357      2   68         16        1
..       ...    ...     ...    ...  ...        ...      ...
395   12.096   4100     307      3   32         13        1
396   13.364   3838     296      5   65         17        0
397   57.872   4171     321      5   67         12        1
398   37.728   2525     192      1   44         13        1
399   18.701   5524     415      5   64          7        0

[400 rows x 7 columns]>
(400,)


In [41]:
my_logreg = LogisticRegression(max_iter=400)

mse_list = cross_val_score(my_linreg, X, y, cv=10, scoring='neg_mean_squared_error')

print(mse_list)

[-23646.90415343 -32003.04401232 -35462.64435619 -37327.60719635
 -14341.32205939 -33628.37104224 -31631.99317834 -12491.00334951
 -20749.61212175 -23204.94743459]


In [43]:
mse_list_positive = -mse_list

rmse_list = np.sqrt(mse_list_positive)
print(rmse_list)

print(rmse_list.mean())

[153.7754992  178.89394627 188.31527914 193.20353826 119.75525901
 183.38039983 177.85385343 111.76315739 144.04725656 152.331702  ]
160.33198910744133
